src/wp-includes/script-loader.php

```diff
@@ -2426,13 +2426,39 @@ function wp_enqueue_registered_block_scripts_and_styles() {
 function enqueue_block_styles_assets() {
 	$block_styles = WP_Block_Styles_Registry::get_instance()->get_all_registered();
 
-	foreach ( $block_styles as $styles ) {
+	foreach ( $block_styles as $block_name => $styles ) {
 		foreach ( $styles as $style_properties ) {
 			if ( isset( $style_properties['style_handle'] ) ) {
-				wp_enqueue_style( $style_properties['style_handle'] );
+
+				// If the site loads separate styles per-block, enqueue the stylesheet on render.
+				if ( wp_should_load_separate_core_block_assets() ) {
+					add_filter(
+						'render_block',
+						function( $html, $block ) use ( $style_properties ) {
+							wp_enqueue_style( $style_properties['style_handle'] );
```

Comment:
It's similar to how the `render_block` works internally:
https://github.com/WordPress/wordpress-develop/blob/bb389cda941be933ced414f5d9406c170e228f01/src/wp-includes/class-wp-block.php#L228-L234

I'm wondering if we should add a check in both places whether the asset was already enqueued.

Reply:
Even in case they get double-enqueued, they will only be printed once so I don't think that will be an issue...